File: bench/microbench.lua
-- Lua microbench by kimio kuramitsu

function testSimpleLoop()
	for i = 0, 100000000 do
	end
	return i
end

function testLocalVariable()
	y = 0
	for i = 0, 100000000 do
		y = y + 1
	end
	return y
end

global_x = 0
global_y = 0
global_z = 0

function testGlobalVariable()
	for i = 0, 100000000 do
		global_y = global_y + 1
	end
	return global_y
end

function func0()
end

function testFunctionCall()
	i = 0
	for i = 0, 100000000 do
		func0()
	end
	return i
end

function func1()
	return 1
end

function testFunctionReturn()
	res = 0
	for i = 0, 100000000 do
		res = func1()
	end
	return res
end

function testFloatOperation()
	f = 0.0
	for i = 0, 100000000 do
		f = f + 0.1
	end
	return f
end

Dim = {}
function Dim.new()
	local instance = {
		x = 0,
		y = 0,
		z = 0
	}
	return instance
end

--

function testFieldVariable()
	d = Dim.new()
	for i = 0, 100000000 do
		d.y = d.y + 1
	end
	return d
end


function run()
	s = os.clock() * 1000
	testSimpleLoop()
	period = os.clock() * 1000 - s
	score = period
	print("SimpleLoop[ms]: ", "period", string.format("%d", period), "score", string.format("%d", score))
	delta = period
	--
	s = os.clock() * 1000
	testLocalVariable()
	period = os.clock() * 1000 - s
	score = period - delta
	print("LocalVariable[ms]:", "period", string.format("%d", period), "score", string.format("%d", score))
	--
	s = os.clock() * 1000
	testGlobalVariable()
	period = os.clock() * 1000 - s
	score = period - delta
	print("GlobalVariable[ms]:", "period", string.format("%d", period), "score", string.format("%d", score))
	--
	s = os.clock() * 1000
	testFunctionCall()
	period = os.clock() * 1000 - s
	score = period - delta
	print("FunctionCall[ms]:", "period", string.format("%d", period), "score", string.format("%d", score))
	--
	s = os.clock() * 1000
	testFunctionReturn()
	period = os.clock() * 1000 - s
	score = period - delta
	print("FunctionReturn[ms]:", "period", string.format("%d", period), "score", string.format("%d", score))
	--
	s = os.clock() * 1000
	testFloatOperation()
	period = os.clock() * 1000 - s
	score = period - delta
	print("FloatOperation[ms]:", "period", string.format("%d", period), "score", string.format("%d", score))
	--
	print("Object-Oriented Programming")
	s = os.clock() * 1000
	testFieldVariable()
	period = os.clock() * 1000 - s
	score = period - delta
	print("FieldVariable[ms]:", "period", string.format("%d", period), "score", string.format("%d", score))
end

run()

File: bench/microbench.py
import time, math

def testSimpleLoop():
	i = 0
	while i < 100000000:
		i += 1
	return i

def testLocalVariable():
	i = 0
	y = 0
	while i < 100000000:
		y += 1
		i += 1
	return y

global_x = 0
global_y = 0
global_z = 0

def testGlobalVariable():
	global global_y
	i = 0
	while i < 100000000:
		global_y += 1
		i += 1
	return global_y

def func0():
	pass

def testFunctionCall():
	i = 0
	while i < 100000000:
		func0()
		i += 1
	return i

def func1():
	return 1

def testFunctionReturn():
	i = 0
	y = 0
	while i < 100000000:
		y = func1()
		i += 1
	return y

def testFloatOperation():
	i = 0
	y = 0.0
	while i < 100000000:
		y += 0.2
		i += 1
	return y

class Dim:
	x = 0
	y = 0
	z = 0
	def __init__(self):
		pass

def testFieldVariable():
	d = Dim()
	i = 0
	while i < 100000000:
		d.y += 1
		i += 1
	return d

def testMathFabs():
	i = 0
	f = 1.0
	while i < 100000000:
		i += 1
		f = math.fabs(-1.0)
	return f

def run():
	s = time.time()
	testSimpleLoop()
	period = (time.time() - s) * 1000
	print "SimpleLoop[ms]: ", period, int(period)
	delta = period
	##
	s = time.time()
	testLocalVariable()
	period = ((time.time() - s) * 1000)
	score = period - delta
	print "LocalVariable[ms]:", period, int(score)
	##
	s = time.time()
	testGlobalVariable()
	period = ((time.time() - s) * 1000)
	score = period - delta
	print "GlobalVariable[ms]:", period, int(score)
	##
	s = time.time()
	testFunctionCall()
	period = ((time.time() - s) * 1000)
	score = period - delta
	print "FunctionCall[ms]:", period, int(score)
	##
	s = time.time()
	testFunctionReturn()
	period = ((time.time() - s) * 1000)
	score = period - delta
	print "FunctionReturn[ms]:", period, int(score)
	##
	s = time.time()
	testFloatOperation()
	period = ((time.time() - s) * 1000)
	score = period - delta
	print "FloatOperation[ms]:", period, int(score)
	s = time.time()
	testMathFabs()
	period = ((time.time() - s) * 1000)
	score = period - (delta)
	print "MathFabs[ms]:", period, int(score)
	##
	print "Object-Oriented Programming"
	s = time.time()
	testFieldVariable()
	period = (time.time() - s) * 1000
	score = period - delta
	print "FieldVariable[ms]:", period, int(score)

run()


